contracts/SurplusBeneficiary.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import { SafeERC20, IERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { FullMath } from "@uniswap/v3-core/contracts/libraries/FullMath.sol";
+import { SurplusBeneficiaryStorageV1 } from "./storage/SurplusBeneficiaryStorage.sol";
+import { ISurplusBeneficiary } from "./interface/ISurplusBeneficiary.sol";
+import { IFeeDistributor } from "./interface/IFeeDistributor.sol";
+
+contract SurplusBeneficiary is ISurplusBeneficiary, ReentrancyGuard, Ownable, SurplusBeneficiaryStorageV1 {
+    using Address for address;
+    using SafeMath for uint256;
+
+    constructor(
+        address tokenArg,
+        address feeDistributorArg,
+        address treasuryArg,
+        uint24 treasuryPercentageArg
+    ) ReentrancyGuard() Ownable() {
+        // SB_TANC: token is not a contract
+        require(tokenArg.isContract(), "SB_TANC");
+        _token = tokenArg;
+
+        setFeeDistributor(feeDistributorArg);
+        setTreasury(treasuryArg);
+        setTreasuryPercentage(treasuryPercentageArg);
+
+        IERC20(_token).approve(feeDistributorArg, type(uint256).max);
```

Comment:
dispatch will fail after we setFeeDistribotor to a new one if treasuryPercentage is not 100%
should move to setFeeDistributor or dispatch

Reply:
will do. Nice catch!